expr/plan.go

```diff
@@ -203,159 +117,99 @@ func newplanFunc(e *expr, fn Func, from, to uint32, stable bool, reqs []Req) ([]
 	//   might be dynamically typed. e.g. movingAvg returns 1..N series depending on how many it got as input
 
 	// first validate the mandatory args
-
-	j := 0 // pos in args of next given arg to process
-	for _, argExp := range argsExp {
-		if len(e.args) <= j {
+	pos := 0    // pos in args of next given arg to process
+	cutoff := 0 // marks the index of the first optional point (if any)
+	var argExp Arg
+	for cutoff, argExp = range argsExp {
+		if argExp.Optional() {
+			break
+		}
+		if len(e.args) <= pos {
```

Comment:
that case can be made, however, right now there is a symmetry with the section below (which loops over the args again but the optional ones) where this condition is not an error. i like that you can look at the two loops and compare the behavior, and clearly see that for one this is an error and for the other it's not